src/schoology-extractor/schoology_extractor/usage_analytics_facade.py

```diff
@@ -3,34 +3,75 @@
 
+from .helpers import csv_reader
+from .helpers import sync
 from .mapping import usage_analytics as usageMap
 
+logger = logging.getLogger(__name__)
 
-def get_system_activities(usage_df: pd.DataFrame) -> pd.DataFrame:
-
-    # TODO: instead of accepting the usage data frame as an argument, accept the
-    # input file directory as an argument. Get all file names from the input
-    # directory. For each file, look in SQLite to see if the file has been
-    # processed (this will necessitate creating a new table, of course). Then
-    # for each file, use the csv_reader class to read in a DataFrame. If
-    # there are multiple files then combine them into the same DataFrame.
-    # Psuedocode below might be improved with some logging. Logging here
-    # is probably better than logging inside the csv_reader or usageMap.
-    #
-    # def get_system_activities(input_directory) -> pd.DataFrame.
-    #   output = pd.DataFrame()
-    #   for file in os.scandir({only .csv or .csv.gz files}):
-    #       raw_data = csv_reader.load_data_frame(file)
-    #       mapped_data = usageMap.map_to_udm(raw_data)
-    #       output.append(mapped_data)
-    #
-    #   # If reports have overlapping dates then we'll have duplicates
-    #   output.drop_duplicates(inplace=True)
-    #
-    #   return output
-
-    return usageMap.map_to_udm(usage_df)
+
+@dataclass
+class UsageAnalyticsFacade:
+    """
+    Provides business logic to orchestrate processing of data from the LMS(System
+    Activities report CSV) and reshape it into the Ed-Fi LMS Unified Data Model
+    (UDM).
+
+    Parameters
+    ----------
+    db_engine : sqlalchemy.engine.base.Engine
+        Database connectivity for sync process
+    """
+    db_engine: sqlalchemy.engine.base.Engine
+
+    @property
+    def _db_engine(self):
+        assert isinstance(self.db_engine, sqlalchemy.engine.base.Engine)
+        return self.db_engine
+
+    def get_system_activities(self, usage_input_dir: str) -> pd.DataFrame:
+        """
+        Processes the .csv or .gz files from the input directory.
+
+        Parameters
+        ----------
+        usage_input_dir: str
+            Directory where the System Activities reports are stored
+
+        Returns
+        -------
+        pd.DataFrame
+            Data reshaped into the Ed-Fi LMS Unified Data Model(UDM)
+        """
+
+        output = pd.DataFrame()
+        logger.debug(f"Processing usage_analytics files: loading files from {usage_input_dir}")
+        for file in os.scandir(usage_input_dir):
+            # It is not expected to have anything different from .gz or .csv
+            # in case there's something different, this method will throw an
+            # exception
+            logger.debug(f"Processing usage_analytics files: processing {file.name}")
+            row_data = csv_reader.load_data_frame(file.path)
+            mapped_row = pd.DataFrame()
+
+            if not sync.usage_file_is_processed(file.name, self._db_engine):

```

Comment:
Shouldn't the check to see if the file has been processed be run before reading the file? it might be nice to avoid the overhead of parsing the input file if we don't need the results.

Reply:
✔️